pallets/identity/src/keys.rs

```diff
@@ -244,6 +244,84 @@ impl<T: Config> Module<T> {
         Ok(())
     }
 
+    /// Accepts a primary key rotation.
+    /// Differs from accept_primary_key_rotation in that it directly swaps a primary key with a secondary key,
+    /// instead of replacing the primary key with an unlinked key, so it requires no authorization.
+    crate fn direct_primary_key_rotation(
+        origin: T::Origin,
+        rotation_auth_id: u64,
+        optional_cdd_auth_id: Option<u64>,
+    ) -> DispatchResult {
+        let new_primary_key = ensure_signed(origin)?;
+        let new_primary_key_signer = Signatory::Account(new_primary_key.clone());
+        Self::accept_auth_with(
+            &new_primary_key_signer,
+            rotation_auth_id,
+            |data, target_did| {
+                let perms = extract_auth!(data, DirectRotatePrimaryKey(p));
+                // Ensure length limits are enforced in `perms`.
+                Self::ensure_perms_length_limited(&perms)?;
+
+                // Accept authorization from CDD service provider.
+                if Self::cdd_auth_for_primary_key_rotation() {
+                    let auth_id = optional_cdd_auth_id
+                        .ok_or_else(|| Error::<T>::InvalidAuthorizationFromCddProvider)?;
+
+                    let signer = Signatory::Account(new_primary_key.clone());
+                    Self::accept_auth_with(&signer, auth_id, |data, auth_by| {
+                        let attestation_for_did = extract_auth!(data, AttestPrimaryKeyRotation(a));
+                        // Attestor must be a CDD service provider.
+                        ensure!(
+                            T::CddServiceProviders::is_member(&auth_by),
+                            Error::<T>::NotCddProviderAttestation
+                        );
+                        // Ensure authorizations are for the same DID.
+                        ensure!(
+                            target_did == attestation_for_did,
+                            Error::<T>::AuthorizationsNotForSameDids
+                        );
+                        Ok(())
+                    })?;
+                }
+
+                DidRecords::<T>::mutate(target_did, |record| {
+                    // Ensure the new primary key is a secondary key.
+                    ensure!(
+                        record
+                            .secondary_keys
+                            .iter()
+                            .any(|si| si.signer == new_primary_key_signer),
+                        Error::<T>::NotASigner
+                    );
+                    let old_primary_key = record.primary_key.clone();
+
+                    // Remove old secondary key
+                    let removed_signers = vec![new_primary_key_signer.clone()];
+                    record.remove_secondary_keys(&removed_signers);
+                    Self::deposit_event(RawEvent::SecondaryKeysRemoved(
+                        target_did,
+                        removed_signers,
+                    ));
```

Comment:
Is this ensure signer check really necessary? Why would we not want to allow a new key be added as a Primary and at the same time make the old key a secondary? If the primary key was able to call it directly without creating an authorization for the secondary key to accept I'd understand the need to restrict to existing an secondary key but as this still requires both parties to accept I believe this check is not required.

I can think of 4 cases to be considered for primary key rotation:
1. Link new unlinked primary and unlink old primary (existing rotate primary key method)
2. Link existing secondary as primary and swap old primary to secondary (this method)
3. Link new unlinked primary and swap old primary to secondary
4. Link existing secondary as primary and unlink old primary

Both 2 & 3 could be covered by this new function if the above ensure signer check was removed and the operation to remove the old secondary key was called only if the new key was an existing signer. One further check to ensure the new primary key is not linked to any DID before joining as primary would also need to be added in this case. 

(Option 4 could arguably also be covered with a modification to the existing AcceptPrimaryKey function.)

Reply:
The goal for this change was just to cover case 2.  The current `RotatePrimaryKey` method was a bad name choice, since most people think it should allow "swapping" a secondary key with their primary key.

I don't think we want to support case 4, the reason is that unlinking the old primary key often leave some POLYX balance unuseable (can't transfer it) since the old key will not have an identity.

Not sure if we will support case 3.

Reply:
@Neopallium Not covering case 3 would be a missed opportunity, in my opinion, and I don't see why it wouldn't be supported. I think it will be a common occurrence, particularly when people are moving from a browser based wallet to a Ledger. It eliminates the need to first add as a secondary key and as a result would give simplified workflow and a better user experience.

I agree that case 1 and case 4 would likely to be less preferred for the reason you mentioned but also case 1 can potentially leave you with a primary key with no POLYX and all your POLYX in your unlinked account.

Reply:
@F-OBrien We are looking into possibly supporting all 4 cases.